boring/src/ssl/callbacks.rs

```diff
@@ -95,73 +103,90 @@ where
         + Sync
         + Send,
 {
-    unsafe {
-        let ssl = SslRef::from_ptr_mut(ssl);
-        let callback_idx = SslContext::cached_ex_index::<F>();
-
-        let callback = ssl
-            .ssl_context()
-            .ex_data(callback_idx)
-            .expect("BUG: psk callback missing") as *const F;
-        let identity = if identity.is_null() {
-            None
-        } else {
-            Some(CStr::from_ptr(identity).to_bytes())
-        };
-        // Give the callback mutable slices into which it can write the psk.
-        let psk_sl = slice::from_raw_parts_mut(psk as *mut u8, max_psk_len as usize);
-        match (*callback)(ssl, identity, psk_sl) {
-            Ok(psk_len) => psk_len as u32,
-            Err(e) => {
-                e.put();
-                0
-            }
+    // SAFETY: boring provides valid inputs.
+
+    let ssl = unsafe { SslRef::from_ptr_mut(ssl) };
+
+    let identity = if !identity.is_null() {
+        Some(unsafe { CStr::from_ptr(identity) }.to_bytes())
+    } else {
+        None
+    };
+
+    // Give the callback mutable slices into which it can write the psk.
+    let psk_sl = unsafe { slice::from_raw_parts_mut(psk as *mut u8, max_psk_len as usize) };
+
+    let ssl_context = ssl.ssl_context().to_owned();
```

Comment:
I think it worth adding a code comment describing why we call `to_owned` here and in other places. I can easily imagine it being removed during refactoring somewhere in the future.

I think we should expose internal `stored_callback` method on `ssl` that will under the hood obtain strong ref for ssl context and return it wrapped in `#[must_use]` struct along with fn pointer.

This way we somehow will have guarantees that callbacks are obtained in a correct way

Reply:
If we remove the `to_owned` calls, the code just doesn't type check anymore.